### PR TITLE
Feat sounder retry

### PR DESCRIPTION
**Pull request by obejarano**
Added a retry to Sounder's main. Most initial runs fail due to discovery so this simple retry helps a bit

### CODE REVIEW

**Comment by obejarano** [comment]
includes changes to config.cc to restore schedule when using a single reference node

File: CC/Sounder/include/recorder.h [hunk 1]
```diff
@@ -15,6 +15,19 @@
...
 
 namespace Sounder {
+
+class RetryableError : public std::runtime_error {
+public:
+    RetryableError()
+        : std::runtime_error("Receiver could not be setup correctly!")
+    {
+    }
+    RetryableError(const std::string& _message)
```
Comment (asedlmayr) [inline CC/Sounder/include/recorder.h:25]:
```suggestion
    explicit RetryableError(const std::string& _message)
```

File: CC/Sounder/include/recorder.h [hunk 2]
```diff
@@ -15,6 +15,19 @@
...
+        : std::runtime_error(_message)
```
Comment (asedlmayr) [inline CC/Sounder/include/recorder.h:26]:
```suggestion
        : std::runtime_error(message)
```

File: CC/Sounder/main.cc [hunk 3]
```diff
@@ -31,27 +32,41 @@ int main(int argc, char* argv[])
...
+            } catch (const SignalException& e) {
+                std::cerr << "SignalException: " << e.what() << std::endl;
+                ret = EXIT_FAILURE;
+                break;
+
+            } catch (const Sounder::RetryableError& rex) {
+                // Discovery usually fails on the first run, re-try
+                std::cout << "Exception: " << rex.what() << " Re-Try!"
+                          << std::endl;
+                usleep(1e6);
```
Comment (asedlmayr) [inline CC/Sounder/main.cc:62]:
```suggestion
                std::this_thread::sleep_for (std::chrono::seconds(1));
```
Why is the sleep necessary?

File: CC/Sounder/main.cc [hunk 4]
```diff
@@ -31,27 +32,41 @@ int main(int argc, char* argv[])
...
         DataGenerator dg(&config);
         dg.GenerateData(FLAGS_storepath);
     } else {
-        try {
-            SignalHandler signalHandler;
+        int cnt = 0;
+        int maxTry = 2;
+        while (cnt++ < maxTry && ret == EXIT_FAILURE) {
+            try {
+                SignalHandler signalHandler;
```
Comment (asedlmayr) [inline CC/Sounder/main.cc:44]: I would move as much of this outside the main loop as possible.

File: CC/Sounder/main.cc [hunk 5]
```diff
@@ -31,27 +32,41 @@ int main(int argc, char* argv[])
...
+                break;
+
+            } catch (const Sounder::RetryableError& rex) {
+                // Discovery usually fails on the first run, re-try
+                std::cout << "Exception: " << rex.what() << " Re-Try!"
+                          << std::endl;
+                usleep(1e6);
+
+            } catch (const std::exception& exc) {
+                std::cerr << "Re-try exceeded... Program terminated Exception: "
```
Comment (asedlmayr) [inline CC/Sounder/main.cc:65]:
```suggestion
                std::cerr << "Exception Encountered... Program terminated due to "
```

File: CC/Sounder/recorder.cc [hunk 6]
```diff
@@ -63,7 +63,7 @@ Recorder::Recorder(Config* in_cfg, unsigned int core_start)
     } catch (std::exception& e) {
         std::cout << e.what() << '\n';
         gc();
-        throw runtime_error("Error Setting up the Receiver");
+        throw RetryableError();
```
Comment (asedlmayr) [inline CC/Sounder/recorder.cc:66]: It might be good to find the error case 1 level down in receiver.  Then you can rethrow e.

**Review by asedlmayr — COMMENTED** [review]
Added some comments and suggestions.  Nothing of critical importance.

File: CC/Sounder/main.cc (hunk 3 above)
Comment (obejarano) [reply CC/Sounder/main.cc:62]: not really necessary but helps to see what's happening during initialization/failure

Comment (obejarano) [reply CC/Sounder/main.cc]: nevermind, it is necessary... need to wait a bit otherwise it will simply fail again if done right away

**Comment by obejarano** [comment]
tested on Falcon with UE RF3E000392 and two chains (at 3.6GHz):
1) joint BS and UE
2) separate BS and UE

**Review by rdoosty — APPROVED** [review]

**Pull request merged**